### ISSUE TITLE
Make copying resource properties easier

### ISSUE THREAD

**Issue by jkeiser** [chef/chef#4284]
Right now, resources that don't have a default value and don't allow `nil` have to be copied over like this:

```
class Blah < Chef::Resource
  resource_name 'blah'
  property :x, String
end

hi = blah 'hi' do
  x 'boo!'
end
lo = blah 'lo'
lo.x = hi.x if hi.property_is_set?(:x)
```

Otherwise you get a `nil` property warning.

We should consider making `nil` be an official sentinel value for "not set" for any resource that does not specify `nil` as its value. Passing `nil` to the property would set it to `nil` for reals. This means getting rid of the `nil` assignment warning as well.

**Comment by lamont-granquist** [comment]
I really think the syntax should just be:

```ruby
# here I mean that nil is the default/unset value
property :version, [ Integer, nil ]
# here I don't allow nil at all and distinguish between unset and should-be-cleared
property :version, [ Integer, :default, :clear ], default: :default
```

And we should maintain the 'tried to set nil but didn't include nil in the validations' warning, and in Chef 13 should make passing nil set nil for reals.

In other words, I don't think we need to change anything here that we weren't planning on doing.

**Comment by coderanger** [comment]
What should the behavior be if `nil` isn't an allowed value but no default is given? Make that a declaration error?

**Comment by lamont-granquist** [comment]
I'm agnostic.  I'd say leave it so that it raises as a validation error on assignment and let it bake for awhile.

Then when we see how its used, either that becomes a declaration error, or we go down a route of automatically adding the default value to validations (or only do that in the default-is-nil case), or we just leave it alone...

Would people actually correctly want to use the ability to ban setting a property to the default value?  Is that a valid use case?  I sorta feel like not, but I sorta don't want to be that proscriptive...

**Comment by lamont-granquist** [comment]
I guess we could always bake the default into the validation and then have an optional opt-out -- which makes the common case short to type, and the uncommon case longer to type...

```
# setting nil passes validation
property :version, Integer
# setting :unset will pass validation (nil will not)
property :version, Integer, default: :unset
# setting nil will raise a validation exception
property :version, Integer, default_passes_validation: false
```

**Comment by jkeiser** [comment]
@lamont-granquist I do want to make sure the system can tell the difference between a real default value and a sentinel, so that we `property_is_set?`, `reset_property` and patchy resources can do their job easily. My proposal would be almost exactly the same as yours, except I would add an `unset_value` to properties to say what the sentinel value is. It would always have a value, and would default to `nil`. That means that if (for example) if the default is a lazy calculation, the @variable would stay set to `nil`, but when you asked for the value it would get the default (call the lazy calc). Your examples would become:

```
# setting nil passes validation
property :version, Integer
# setting :unset will pass validation (nil will not)
property :version, Integer, unset_value: :unset
# setting nil will raise a validation exception
property :version, Integer, unset_value: :unset
```

These examples could be given defaults (even lazy ones), and the default values would be required to pass all the same rules as the unset_value. Like, `property :version, Integer, unset_value: :unset, default: lazy { calc_default_version }`.

**Comment by jkeiser** [comment]
Most people wouldn't set `unset_value`, but if you do want `nil` to be a thing for you, you would do that.

**Comment by jkeiser** [comment]
Like, anytime nil was a valid value: `property :version, [ Integer, nil ], unset_value: :unset`

**Comment by lamont-granquist** [comment]
how is that different from this:

```
property :version, [ Integer, nil ], default: :unset
```

still think that people who actually want to fix this problem you're concerned about have all the tools they need right now, without adding more complexity to the API...

**Comment by jkeiser** [comment]
The big differences are:

1. There is no way to add a lazy default to a property like that, because `default` has to be used to indicate your sentinel.
2. There is no way to write a patchy property like `mode` if you want to have a default, because there is no way to tell the difference between "user set the mode" and "the mode happens to be set to its default."

We're introducing complexity to the API either way by adding the concept of a sentinel for unset values here; the only question is whether we conflate that concept with default.

**Comment by jkeiser** [comment]
Thinking about it, if we don't require that "whether the property is set"ness is copied over when you copy a property over, you don't need a separate sentinel value. I'll target that (your proposal) first, it makes sense.

**Comment by jkeiser** [comment]
So the basics of this proposal are that, if its default value is constant, and the value is equal to the default value, the property has not been set. This means that if you want to make a patchy resource where `nil` is a valid value, it's on you to make your default something like `:unset`. Yes?

The changes to the system:

1. Instance variables are initialized to their defaults when the resource is created.
2. If an instance variable is set to its default value, `property_is_set?(:property)` will return `false`.
3. `reset_property(:property)` will set the variable to its default value.
4. Default values will not be coerced or validated (right now they are). This seems like a big change honestly, but if we don't do it, you won't be able to set a default to an invalid value.

**Comment by jkeiser** [comment]
We need to decide whether we can do this before Chef 13; implementing these rules across the board means that in most cases, `resource.property(nil)` will set the value to `nil`. To be honest, nearly everything I've seen that actually *does* that, actually seems to mean it--they want to set the value to `nil`. So this might be a net positive rather than negative, and a simple bugfix.

**Comment by jkeiser** [comment]
Hrm, the one thing I don't like about that set of rules is it doesn't fix mode.

`property :mode, Integer, default: 0655`

As before, it matters that we know whether the user said `mode 0655` or said nothing. If the file already exists, it's the difference between setting the mode to 0655, and leaving it alone. We need `mode` to have a different internal value depending on whether the user set it or not.

**Comment by jkeiser** [comment]
For contrast, the current code internally uses the presence or absence of the instance variable to determine if the property is set or not.

**Comment by jkeiser** [comment]
I believe this is fixed by https://github.com/chef/chef/pull/4296 . Here's how these things will work:

- `property(nil)` will only warn if you do not specify a value.
- If you do not specify a default, `nil` bypasses coercion and is a valid value. (This is not a change.)
- If you do specify a default, it will go through coercion and validation. (This is not change.)
- If a property explicitly accepts `nil`, `nil` will go through coercion and validation. (This is not a change.)

The upshot:

1. `property :x`
  - `x(nil)` succeeds, because `nil` is a valid value. `nil` is not coerced if `coerce` is specified.
2. `property :x, String`
  - `x(nil)` succeeds, because `nil` is a valid value meaning "value not set." `nil` is not coerced if `coerce` is specified.
3. `property :x, String, default: 10`
  - `x(nil)` fails, but it's not an issue for copying because the value will never be `nil`.
4. `property :x, String, default: nil`
  - *Fails to compile*, because specifying `nil` as the default means you intended it to be a real value (it's not the special "property does not exist" default anymore) and it does not pass validation. Either remove `default: nil` or add `nil` as a valid value, if it has meaning to you.
5. `property :x, [ String, nil ]`
  - `x(nil)` (and copying `nil`) succeeds because `nil` is a valid value.
6. `property :x, [ String, nil ]`, default: nil
  - `x(nil)` (and copying `nil`) succeeds because `nil` is a valid value. 

In all cases, either copying succeeds, or the resource fails to compile in the first place.  There can be legitimate disagreement about (4), and we should discuss the distinction for Chef 13, but I think this ruleset gets us everything we need to fix this issue.

**Comment by jkeiser** [comment]
The new rules are implemented in https://github.com/chef/chef/pull/4300 and build on top of the no-warnings PR.

**Comment by jkeiser** [comment]
Calling this fixed with #4300. We'll continue to talk about the further Chef 13-ish PR.
